consumer.go

```diff
@@ -0,0 +1,286 @@
+	Name             string `json:"name"`
+	StationName      string `json:"station_name"`
+	ConnectionId     string `json:"connection_id"`
+	ConsumerType     string `json:"consumer_type"`
+	ConsumerGroup    string `json:"consumers_group"`
+	MaxAckTimeMillis int    `json:"max_ack_time_ms"`
+	MaxMsgDeliveries int    `json:"max_msg_deliveries"`
+}
+
+type removeConsumerReq struct {
+	Name        string `json:"name"`
+	StationName string `json:"station_name"`
+}
+
+type ConsumerOpts struct {
+	Name                     string
+	StationName              string
+	ConsumerGroup            string
+	PullIntervalMillis       int
+	BatchSize                int
+	BatchMaxTimeToWaitMillis int
+	MaxAckTimeMillis         int
+	MaxMsgDeliveries         int
+}
+
+func GetDefaultConsumerOptions() ConsumerOpts {
+	return ConsumerOpts{
+		PullIntervalMillis:       1000,
+		BatchSize:                10,
+		BatchMaxTimeToWaitMillis: 5000,
+		MaxAckTimeMillis:         30000,
+		MaxMsgDeliveries:         10,
+	}
+}
+
+type ConsumerOpt func(*ConsumerOpts) error
+
+func (c *Conn) CreateConsumer(stationName, consumerName string, opts ...ConsumerOpt) (*Consumer, error) {
+	defaultOpts := GetDefaultConsumerOptions()
+
+	defaultOpts.Name = consumerName
+	defaultOpts.StationName = stationName
+	defaultOpts.ConsumerGroup = consumerName
+
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt(&defaultOpts); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return defaultOpts.CreateConsumer(c)
+}
+
+func (opts *ConsumerOpts) CreateConsumer(c *Conn) (*Consumer, error) {
+	consumer := Consumer{Name: opts.Name,
+		ConsumerGroup:      opts.ConsumerGroup,
+		PullIntervalMillis: opts.PullIntervalMillis,
+		BatchSize:          opts.BatchSize,
+		MaxAckTimeMillis:   opts.MaxAckTimeMillis,
+		MaxMsgDeliveries:   opts.MaxMsgDeliveries,
+		BatchMaxTimeToWait: time.Duration(opts.BatchMaxTimeToWaitMillis) * time.Millisecond,
+		conn:               c,
+		stationName:        opts.StationName}
+
+	err := c.create(&consumer)
+	if err != nil {
+		return nil, err
+	}
+
+	consumer.firstFetch = true
+	consumer.puller = make(chan *Msg, consumer.BatchSize)
+	consumer.pullerQuit = make(chan struct{}, 1)
+
+	ackWait := time.Duration(consumer.MaxAckTimeMillis) * time.Millisecond
+	subj := consumer.stationName + ".final"
+	durableName := consumer.ConsumerGroup
+	if durableName == "" {
+		durableName = consumer.Name
+	}
+
+	consumer.subscription, err = c.brokerSubscribe(subj, durableName,
+		nats.ManualAck(),
+		nats.AckWait(ackWait),
+		nats.MaxRequestExpires(consumer.BatchMaxTimeToWait),
+		nats.MaxRequestBatch(opts.BatchSize))
+	if err != nil {
+		return nil, err
+	}
+
+	return &consumer, err
+}
+
+func (s *Station) CreateConsumer(name string, opts ...ConsumerOpt) (*Consumer, error) {
+	return s.conn.CreateConsumer(s.Name, name, opts...)
+}
+
+type ConsumeHandler func([]*Msg, error)
+
+func (c *Consumer) Consume(pullInterval time.Duration, handlerFunc ConsumeHandler) {
```

Comment:
pullInterval this parameter is not needed because the user already sent it when creating the consumer